crates/cheatnet/src/execution/syscalls.rs

```diff
@@ -278,24 +277,64 @@ impl CheatableSyscallHandler<'_> {
             response_wrapper.write(vm, &mut self.syscall_handler.syscall_ptr)?;
             return Ok(());
         } else if SyscallSelector::CallContract == selector {
+            let StarknetHint::SystemCall { system: syscall } = hint else {
+                return Err(HintError::CustomHint(
+                    "Test functions are unsupported on starknet.".into(),
+                ));
+            };
+            let initial_syscall_ptr = get_ptr_from_res_operand_unchecked(vm, syscall);
+            self.verify_syscall_ptr(initial_syscall_ptr)?;
+
             // Increment, since the selector was peeked into before
             self.syscall_handler.syscall_ptr += 1;
+
+            if selector != SyscallSelector::Keccak {
+                self.syscall_handler
+                    .increment_syscall_count_by(&selector, 1);
+            }
             return self.execute_syscall(
                 vm,
                 call_contract_syscall,
                 constants::CALL_CONTRACT_GAS_COST,
             );
         } else if SyscallSelector::LibraryCall == selector {
+            let StarknetHint::SystemCall { system: syscall } = hint else {
+                return Err(HintError::CustomHint(
+                    "Test functions are unsupported on starknet.".into(),
+                ));
+            };
+            let initial_syscall_ptr = get_ptr_from_res_operand_unchecked(vm, syscall);
+            self.verify_syscall_ptr(initial_syscall_ptr)?;
+
             // Increment, since the selector was peeked into before
             self.syscall_handler.syscall_ptr += 1;
+
+            if selector != SyscallSelector::Keccak {
+                self.syscall_handler
+                    .increment_syscall_count_by(&selector, 1);
+            }
+
             return self.execute_syscall(
                 vm,
                 library_call_syscall,
                 constants::CALL_CONTRACT_GAS_COST,
             );
         } else if SyscallSelector::Deploy == selector {
+            let StarknetHint::SystemCall { system: syscall } = hint else {
+                return Err(HintError::CustomHint(
+                    "Test functions are unsupported on starknet.".into(),
+                ));
+            };
+            let initial_syscall_ptr = get_ptr_from_res_operand_unchecked(vm, syscall);
+            self.verify_syscall_ptr(initial_syscall_ptr)?;
```

Comment:
There is a lot of repetition here, can we somehow clean this up?

Reply:
I will tackle those in a separate PR https://github.com/orgs/foundry-rs/projects/3/views/2?pane=issue&itemId=40983148 as there is more extensive work to do there.

Reply:
I did some refactors here because of clippy, I still want to clean it up further though

Reply:
https://github.com/foundry-rs/starknet-foundry/pull/680/commits/be56df31c0850bddff991e82c87cf608316dc7ed